blender: toggle for realizing instances when baking

seandillon92 opened the pull request.

Commit:
Scale height scale to be in centimeters.

Commit:
Adding missing meta files

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_support

Commit:
Docs update

Commit:
[skip ci]

Commit:
[skip ci]

Commit:
[skip ci]

Commit:
[skip ci]

Commit:
[skip ci]

Commit:
[skip ci]

Commit:
[skip ci]

Commit:
[skip ci]

Commit:
Moving test files

Commit:
Refactoring

Commit:
Docs update. [skip ci]

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_support

Commit:
Merge branch 'material_support' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking

Commit:
WIP

Commit:
wip

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking

# Conflicts:
#	Plugins~/Src/MeshSyncClientBlender/BlenderSyncSettings.h
#	Plugins~/Src/MeshSyncClientBlender/msblenMaterialsExportHelper.cpp
#	Plugins~/Src/MeshSyncClientBlender/msblenMaterialsExportHelper.h
#	Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_common.py

Commit:
wip

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking

# Conflicts:
#	Plugins~/Src/MeshSyncClientBlender/msblenBinder.h

Commit:
WIP

Commit:
WIP material baking

Commit:
Fixes and reverting unwanted stuff

Commit:
Reverting not needed changes

Commit:
Revert

Commit:
More reverting

Commit:
Format fix

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking

Commit:
Adding channel settings

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking

# Conflicts:
#	Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_common.py

Commit:
Adding comment

Commit:
format fix

Commit:
Removing unused code

Commit:
Starting work to optionally bake inputs to connected bsdf

Commit:
Disabling channels we cannot bake properly yet.
Updating docs.

Commit:
Docs update

Commit:
Docs update

Commit:
Merge branch 'material_baking' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking_inputs

Commit:
wip

Commit:
Support to bake bsdf inputs directly with falllback.
Bug fixes

Commit:
Channel mapping fixes to support all shader types.
Adding test file.

Commit:
Bug fixes.
Changing blender export to send roughness maps.

Commit:
Removing test code

Commit:
Making settings revertable.

Commit:
Ungroup material node groups if needed

Commit:
Refactoring to put all bake settings into their own class.
Fix for baking objects in hidden collections.

Commit:
Fix to restore selection after baking.
Make objects single user before applying modifiers.

Commit:
Updating test file

Commit:
Updating docs.
Reuse node socket inputs instead of baking the same input multiple times.

Commit:
Making execution modal

Commit:
Modal run improvements.
Adding option to run modal or not.
Adding remaining time estimate.

Commit:
Improved remaining time calculation.

Commit:
typo fix

Commit:
progress message update

Commit:
Adding comments

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking_ungrouping

# Conflicts:
#	Plugins~/Src/MeshSyncClientBlender/msblenBinder.cpp
#	Plugins~/Src/MeshSyncClientBlender/python/3.3.0/unity_mesh_sync.py
#	Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_common.py

Commit:
Updating test file

Commit:
WIP: Use texel ratio calculation to determine baked texture size.

Commit:
Adding texel density setting for baked texture size.

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking_ungrouping

# Conflicts:
#	Plugins~/Src/MeshSyncClientBlender/python/2.90.1/unity_mesh_sync.py
#	Plugins~/Src/MeshSyncClientBlender/python/2.91.2/unity_mesh_sync.py
#	Plugins~/Src/MeshSyncClientBlender/python/2.92.0/unity_mesh_sync.py
#	Plugins~/Src/MeshSyncClientBlender/python/2.93.7/unity_mesh_sync.py
#	Plugins~/Src/MeshSyncClientBlender/python/3.0.1/unity_mesh_sync.py
#	Plugins~/Src/MeshSyncClientBlender/python/3.1.0/unity_mesh_sync.py
#	Plugins~/Src/MeshSyncClientBlender/python/3.2.0/unity_mesh_sync.py
#	Plugins~/Src/MeshSyncClientBlender/python/3.3.0/unity_mesh_sync.py
#	Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_common.py
#	Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_preferences.py

Commit:
Reducing nesting

Commit:
Refactoring

Commit:
Adding comments

Commit:
Adding check to ensure GPU is available before setting it.

Commit:
Updating docs

Commit:
Updating docs.

Commit:
Removing old comment

Commit:
Improved material output detection

Commit:
Adding test for texel density

Commit:
Removing dead code

Commit:
Removing unused stuff

Commit:
Merge branch 'material_baking_ungrouping' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking_texel_ratio

Commit:
Updating docs

Commit:
Updating docs

Commit:
updating docs

Commit:
Updating docs

Commit:
updating docs

Commit:
updating docs

Commit:
Updating docs

Commit:
WIP

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking_ungrouping

Commit:
Adding 3.4 support

Commit:
Register default settings for all blender versions

Commit:
Updating docs

Commit:
Updating log level

Commit:
Adding message for escape event

Commit:
Merge branch 'material_baking_ungrouping' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking_ao

Commit:
[skip-ci] Updated doc formatting for Generate UV modes

Commit:
[skip-ci] Fix "If Needed" formatting in Generate UV options doc

Commit:
[skip-ci] formatting fix for Apply modifiers option

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking_ao

Commit:
setting log level

Commit:
Adding a way to bake ambient occlusion.

Commit:
Removing dead code

Commit:
Updating description

Commit:
Extracting method.

Commit:
Removing dead code

Commit:
Refactoring

Commit:
Refactoring

Commit:
Removing unused method parameter

Commit:
Disabling exceptions for release

Commit:
Merge branch 'material_baking_ungrouping' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking_texel_ratio

# Conflicts:
#	Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py

Commit:
Updating docs

Commit:
Updating docs

Commit:
Updating docs

Commit:
Updating docs

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking_texel_ratio

# Conflicts:
#	Documentation~/Blender.md
#	Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py
#	Plugins~/TestAssets/Test scenes/bake_test_1.blend

Commit:
Merge fix

Commit:
Merge branch 'material_baking_texel_ratio' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking_ao

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_baking_ao

# Conflicts:
#	Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py

Commit:
Fix for baking objects in excluded collections.

Commit:
Reverting debug change

Commit:
Fix to ignore non-mesh objects when prebaking.

Commit:
blender-feat: Adding handling for value nodes (#335)

Commit:
Fix for weight calculation.
Fix for empty material slots baking.
Fix for materials crash.

Commit:
Apply modifiers before any baking logic to ensure materials and geometry do not change.
Fix to apply modifiers on objects in disabled collections.

Commit:
Merge branch 'material_baking_ao' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_fixes_2

Commit:
Fix for material indices in blender 3.4

Commit:
Baking support for all versions before 3.3.0

Commit:
Fix to keep unbaked bsdf inputs connected.

Commit:
Merge branch 'material_fixes_2' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_fixes_2

Commit:
visibility fix

Commit:
Adding check for tiled images

Commit:
Refactoring

Commit:
Fix for older blender versions

Commit:
Don't bake RGB and Value nodes, meshsync can export those now.

Commit:
toggle to add realize instance nodes before baking

schinkowski commented on Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:
```diff
@@ -369,6 +373,13 @@ def preBakeObject(self, obj):
                 bpy.ops.object.make_single_user(type='SELECTED_OBJECTS', obdata=True)
                 for mod in obj.modifiers[:]:
+                    if bakeSettings.realize_instances and mod.type == "NODES":
+                        output = mod.node_group.nodes["Group Output"]
```
That name is not guaranteed to exist, users can rename nodes. I think you'll need to loop through the nodes and find nodes with the output type and probably check which is the active output in case there are more than 1. I'm doing something similar in `findMaterialOutputNode` for the material graph.

seandillon92 replied:
I changed it to use the type. I could not find a reliable way to get the right group output, since active would point to the last selected output node or -1 if none has been selected yet. So instead, I just iterate over all output nodes and add the realize instances, assuming it will be a small number.

Commit:
Merge branch 'dev' of https://github.com/Unity-Technologies/MeshSyncDCCPlugins into material_fixes_2

# Conflicts:
#	Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py

Commit:
merge fix

Commit:
Updating docs

Commit:
UI improvements

Commit:
Changing default UVs setting to ensure it always works

Commit:
Adding default for baking folder

Commit:
UI improvement.
Fix for modifiers not being applied when needed.

Commit:
Merge branch 'material_fixes_2' into realize-instances-toggle

Commit:
improved code to work with renames of output node

Commit:
iterate over all group output nodes

schinkowski commented on Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:
```diff
@@ -347,6 +351,19 @@ def incrementProgress(self, context, message, obj=None, reset=False):
+        outputs = [x for x in nodes if x.type == "GROUP_OUTPUT"]
+
+        for output in outputs:
+            if len(output.inputs) == 0 or len(output.inputs[0].links) == 0:
```
I think instead of checking the number of inputs, the 'is_active_output' flag should be used.

seandillon92 replied:
I tried that but noticed that is_active_input is true for the node that is selected last. So if have 2 outputs nodes and only one has links but I select the empty one, the bake will not include the instances. 

My solution is to simply try and add the realize instance node if possible, that is the output node has at least one input with at least one link.

schinkowski replied:
Yeah the selected node should be the active one, that's how it works for materials as well. Would there be any harm to just add the realize instances node before every output?

Commit:
fix for recursive yield function

Commit:
Ensure baking happens to correct UV map.
Fix to handle unselectable collections.
Ensure bake happens to correct normal space.

Commit:
Merge branch 'material_fixes_2' into realize-instances-toggle

Commit:
avoid applying particle system modifiers

Commit:
Removing forgotten line

Commit:
Increasing samples to remove noise

Commit:
Excluding armature modifiers to avoid breaking animations

Commit:
Merge remote-tracking branch 'origin/material_fixes_2' into realize-instances-toggle

Commit:
handled merge conflicts with dev
